akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```diff
@@ -354,4 +389,51 @@ object HttpEntity {
     def getTrailerHeaders: JIterable[jm.HttpHeader] = trailer.asJava
   }
   object LastChunk extends LastChunk("", Nil)
+
+  /**
+   * Turns the given source into one that respects the `withSizeLimit` calls when used as a parameter
+   * to entity constructors.
+   */
+  def limitable[Out: SizeOfInBytes, Mat](source: Source[Out, Mat]): Source[Out, Mat] =
+    source.transform { () ⇒
+      new PushStage[Out, Out] {
+        val sizeOf = implicitly[SizeOfInBytes[Out]]
+        var maxBytes = -1L
+        var bytesLeft = Long.MaxValue
+
+        override def preStart(ctx: LifecycleContext) =
+          ctx.attributes.getFirst[SizeLimit] match {
+            case Some(SizeLimit(bytes, cl @ Some(contentLength))) ⇒
+              if (contentLength > bytes) throw EntityStreamSizeException(bytes, cl)
+            // else we still count but never throw an error
+            case Some(SizeLimit(bytes, None)) ⇒
+              maxBytes = bytes
+              bytesLeft = bytes
+            case None ⇒
+          }
+
+        def onPush(elem: Out, ctx: stage.Context[Out]): stage.SyncDirective = {
+          bytesLeft -= sizeOf(elem)
+          if (bytesLeft >= 0) ctx.push(elem)
+          else ctx.fail(EntityStreamSizeException(maxBytes))
+        }
+      }
+    }
```

Comment:
I'm looking for a simple, low-overhead (both in terms of LOCs and runtime overhead) way to mark the transform stage with a name, that's all. Currently there doesn't appear to be one, unless I widen the `private[stream]` marker on `PushPullGraphStage` to `private[akka]`. If you think that that's the way to go I'm happy to make this change.

Otherwise I think I have to do something like

```scala
source.via(Flow[Out].transform { () ⇒
  new PushStage[Out, Out] {
    ...
  }
}.named("limit"))
```

which works but isn't really "nice", neither from a source nor from a runtime perspective (2 nested `via` constructs).